cardano-api/src/Cardano/Api/Value.hs

```diff
@@ -115,6 +131,98 @@ instance Semigroup Value where
 instance Monoid Value where
   mempty = Value Map.empty
 
+instance ToJSON Value where
+  toJSON = toJSON . unflatten
+
+unflatten :: Value -> AssetsBundle
+unflatten (Value flatMap) = AssetsBundle $ Map.foldlWithKey folder [] flatMap
+ where
+  folder
+    :: [AssetIdBundle]
+    -> AssetId
+    -> Quantity
+    -> [AssetIdBundle]
+  folder acc (AssetId pid aName) q = AssetIdBundle pid (Map.singleton aName q) : acc
+  folder acc AdaAssetId q = AdaAsset q : acc
+
+instance FromJSON Value where
+  parseJSON jv = flatten <$> parseJSON jv
+
+flatten :: AssetsBundle -> Value
+flatten (AssetsBundle bundles) = foldl folder (Value Map.empty) bundles
+ where
+  folder :: Value -> AssetIdBundle -> Value
+  folder acc (AssetIdBundle pid assetNameQuantityMap) =
+   acc <> Value (Map.mapKeys (\aname -> AssetId pid aname) assetNameQuantityMap)
+  folder acc (AdaAsset q) = acc <> Value (Map.singleton AdaAssetId q)
+
+-- | Intermediate representation used in the JSON parsing/rendering of 'Value'
+newtype AssetsBundle = AssetsBundle [AssetIdBundle]
+                       deriving (Eq, Ord, Show)
+
+data AssetIdBundle = AssetIdBundle PolicyId (Map AssetName Quantity)
+                   | AdaAsset Quantity
+                   deriving (Eq, Ord, Show)
+
+instance ToJSON AssetsBundle where
+  toJSON (AssetsBundle bundles) = object $ map toPair bundles
+    where
+     toPair :: AssetIdBundle -> (Text, Aeson.Value)
+     toPair (AdaAsset q) = ("lovelace", toJSON q)
+     toPair (AssetIdBundle pid assets) = (renderPolicyId pid, toJSON assets)
+
+     renderPolicyId :: PolicyId -> Text
+     renderPolicyId (PolicyId sh) = Text.decodeUtf8 (serialiseToRawBytesHex sh)
+
+instance ToJSON Quantity where
+  toJSON (Quantity q) = toJSON q
+
+instance FromJSON Quantity where
+  parseJSON = withScientific "MultiAssetQuantity" $ \sci ->
+    case Scientific.floatingOrInteger sci :: Either Double Integer of
+      Left  d -> fail $ "Expected an integer but got: " <> show d
+      Right n -> return $ Quantity n
+
+instance ToJSON AssetName where
+  toJSON (AssetName an) = Aeson.String $ Text.decodeUtf8 an
+
+instance FromJSON AssetName where
+  parseJSON = withText "MultiAssetName"
+                $ \text -> return . AssetName $ Text.encodeUtf8 text
+
+instance ToJSONKey AssetName where
+  toJSONKey = toJSONKeyText
+                $ \(AssetName asset) -> Text.decodeUtf8 asset
+
+instance FromJSONKey AssetName where
+  fromJSONKey = FromJSONKeyText
+                  $ \assetName -> AssetName $ Text.encodeUtf8 assetName
+
+instance FromJSON AssetsBundle where
+  parseJSON = withObject "AssetsBundle" $ \obj ->
+                AssetsBundle <$> HashMap.foldlWithKey' folder (return []) obj
+   where
+     folder
+       :: Parser [AssetIdBundle]
+       -> Text
+       -> Aeson.Value
+       -> Parser [AssetIdBundle]
+     folder acc pid asset = do
+       accum <- acc
+       assetIdBundle <- parsePid pid asset
+       return $ assetIdBundle : accum
```

Comment:
Can this not also be done in a bulk style now? The AssetsBundle type matches 1:1 with the JSON structure we need. The `toJSON` uses a bulk style.